### DISCUSSION TITLE
Pytrees: `defaultdict` key order is significant?

### DISCUSSION THREAD

**Discussion by cherrywoods** [jax-ml/jax#22987] [General] [answered]
Hi,
I have a bit of a niche question: When looking through the base Pytree registry definitions in `jax/_src/tree_util.py`, I stumbled across the definition for `defaultdicts` in this line: https://github.com/google/jax/blob/9beb4f14741679eef8f0cf8af22c04b4509e7448/jax/_src/tree_util.py#L794

For regular `dicts`, the keys are sorted so that dictionaries with different insertion order still correspond to the same pytree. However, for `defaultdicts` the keys are not sorted. I understand why the keys are not sorted for `OrderedDict`, but would have expected them to get sorted for `defaultdict`. 
I am just curious here whether I'm missing a detail of Python or, otherwise, what the rationale is for not sorting the keys of `defaultdicts`.

**Comment by yashk2810** [comment, marked as answer]
Good point, I think we should sort it. @hawkinsp WDYT?

**Reply by yashk2810** [reply]
@jakevdp too

**Reply by jakevdp** [reply]
Yes, we should sort keys on lowering here.
